Propagated Sampling Rates (#11912)
getsentry/sentry-docs: develop-docs/sdk/telemetry/traces/index.mdx

### `tracesSampleRate`

This should be a float/double between `0.0` and `1.0` (inclusive) and represents the percentage chance that any given transaction will be sent to Sentry. So, barring [outside influence](#sampling), `0.0` is a 0% chance (none will be sent) and `1.0` is a 100% chance (all will be sent). This rate applies equally to all transactions; in other words, each transaction should have the same random chance of ending up with `sampled = true`, equal to the `tracesSampleRate`.

See more about how sampling should be performed [below](#sampling).

### `tracesSampler`

This should be a callback, called when a transaction is started, which will be given a `samplingContext` object and which should return a sample rate between `0.0` and `1.0` _for the transaction in question_. This sample rate should behave the same way as the `tracesSampleRate` above, with the difference that it only applies to the newly-created transaction, such that different transactions can be sampled at different rates. Returning `0.0` should force the transaction to be dropped (set to `sampled = false`) and returning `1.0` should force the transaction to be sent (set `sampled = true`).

> Historically, the `tracesSampler` callback could have also returned a boolean to force a sampling decision (with `false` equivalent to `0.0` and `true` equivalent to `1.0`). This behavior is now **deprecated** and should be removed from all SDKs.

Reviewer: What would be the benefit of doing so? 

On first glance, I see a couple of downsides in having booleans in the API:
1. It creates multiple overloads in the function signature or a polymorphic/mixed return type
2. The semantics of `1.0` (100%) for extrapolation are inherently clear, but for `return True` not so much. My expectation would be that we "inherit" the parent extrapolation factor, but in reality we apply a hidden 100% sample rate. 
3. It's multiple ways to do the same thing.

</Alert>

### `strictTracePropagation`

> This option disables trace continuation from unknown 3rd party services that happen to be instrumented by a Sentry SDK.

Reviewer 2: ```suggestion
This should be a boolean value. Default is `false`. This option controls trace continuation from unknown 3rd party services that happen to be instrumented by a Sentry SDK.
```

> If the SDK can parse an org ID from the configured DSN, this value must be propagated as a baggage entry with the key `sentry-org`. Given a DSN of `https://1234@o1.ingest.us.sentry.io/1`, the org ID is `1`, based on `o1`.

Reviewer 2: ```suggestion
If the SDK can parse an org ID from the configured DSN, it must be propagated as a baggage entry with the key `sentry-org`. Given a DSN of `https://1234@o1.ingest.us.sentry.io/1`, the org ID is `1`, based on `o1`.
```

On incoming traces, the SDK must compare the `sentry-org` baggage value against its own parsed value from the DSN. Only if both match, the trace is continued. If there is no match, neither the trace ID, the parent sampling decision nor the baggage should be taken into account.

> The SDK should behave like it is the head of trace in this case.

Reviewer 2: This sentence I do not understand. Can you please reword this?

This behavior can be disabled by setting `strictTracePropagation: false` in the SDK init call.
Initially, SDKs should introduce the this option with a default values of `false`.

> Once the majority of SDKs introduced this option, we will make it opt-out in a major version.

Reviewer 2: ```suggestion
Once the majority of SDKs introduced this option, we will change the default value to `true` (in a major version bump) making it opt-out.
```

</Alert>

> ### `strictTracePropagation`

Author: FMI, we want to consider to auto discard traces where the incoming trace contains an org that does not match the org configured in the SDK.

The SDK should behave like it is the head of trace in this case, and not consider any propagted values.

This behavior can be disabled by setting `strictTracePropagation: false` in the SDK init call.
Initially, SDKs must introduce the this option with a default value of `false`.
Once the majority of SDKs introduced this option, we will change the default value to `true` (in a major version bump) making it opt-out.

Regardless of `strictTracePropagation` being set to `true` or `false`, if the SDK is either configured with a `org` or was able to parse the value from the DSN, incoming traces containing an `org` value in the baggage that does not match the one from the receiving SDK, the trace is not continuned.

Examples:

baggage: `sentry-org: 1`, SDK config: `org: 1, strictTracePropagation: false` -> continue trace
baggage: `sentry-org: none`, SDK config: `org: 1, strictTracePropagation: false` -> continue trace
baggage: `sentry-org: 1`, SDK config: `org: none, strictTracePropagation: false` -> continue trace
baggage: `sentry-org: none`, SDK config: `org: none, strictTracePropagation: false` -> continue trace

> baggage: `sentry-org: 1`, SDK config: `org: 2, strictTracePropagation: false` -> drop trace

Reviewer 3: When reading `drop trace` I might consider actually dropping/discarding the trace, but shouldn't the SDK start a new trace and discontinue the trace?

Author: ~~drop trace~~ -> start new trace

## Sampling

> Each transaction has a _sampling decision_, that is, a boolean which declares whether or not it should be sent to Sentry. This should be set exactly once during a transaction's lifetime, and should be stored in an internal `sampled` boolean.

Author: I would like to keep it as is, as it's kinda important to only do it once, and i wanted to add some emphasis.

A transaction's sampling decision should be passed to all of its children, including across service boundaries. This can be accomplished in the `startChild` method for same-service children and using the `senry-trace` header for children in a different service.

### Propagated Sampling Rates

> To increase the chance of capturing complete traces when users return a new sample rate from the `tracesSampler` in backend services, we propagate the generated random value used by the SDK for computing the sampling decision instead of creating a new random value in every service. Therefore, across a trace every SDK uses the same random value.

Reviewer 2: ```suggestion
To increase the chance of capturing complete traces when users return a new sample rate from the `tracesSampler` in backend services, we propagate the generated random value used by the SDK for making the sampling decision instead of creating a new random value in every service. Therefore, across a trace every SDK uses the same random value.
```

The behavior of the static `tracesSampleRate` without the use of `tracesSampler` does not change. We continue to fully inherit sampling decisions for propagated traces and create a new one for started traces. In the future, we might change the default behavior of `tracesSampleRate`, too.

> Once the SDK starts a new trace, `sentry-sample_rand` is filled with a truly random number. This also applies when the `tracesSampleRate` is set to `1.0`. If the SDK receives an incoming trace, containing a `sentry-sample_rand` entry in the baggage, this value should overwrite the currently stored number. For incoming traces without a `sentry-sample_rand` baggage entry (from old SDKs), the SDK inserts a new truly random number on-the-fly.

Reviewer 2: ```suggestion
When a new trace is started by the SDK, `sentry-sample_rand` is filled with a truly random number. This also applies when the `tracesSampleRate` is set to `1.0`. If the SDK receives an incoming trace, containing a `sentry-sample_rand` entry in the baggage, this value should overwrite the currently stored number. For incoming traces without a `sentry-sample_rand` baggage entry (from old SDKs), the SDK inserts a new truly random number on-the-fly.
```

The SDK should exclusively use the stored random number and no longer use `math.random()` or similar anywhere else in the tracing related code base.

- When the `tracesSampler` is invoked, this also applies to the return value of traces sampler. i.e. `trace["sentry-sample_rand"] < tracesSampler(context)`
- Otherwise, when the SDK is the head of a trace, this applies to sample decisions based on `tracesSampleRate` , i.e. `trace["sentry-sample_rand"] < config.tracesSampleRate`

> When using a `tracesSampler`, the most correct way to inherit parent sampling decisions is now to return the parent's sample rate instead of the decision as float (1.0). This way, we can still extrapolate counts correctly.

Reviewer 2: ```suggestion
When using a `tracesSampler`, the correct way to inherit parent sampling decisions is now to return the parent's sample rate instead of the decision as float (1.0). This way, we can still extrapolate counts correctly.
```

A transaction's sampling decision should be passed to all of its children, including across service boundaries. This can be accomplished in the `startChild` method for same-service children and using the `senry-trace` header for children in a different service.

### Propagated Random Value

To increase the chance of capturing complete traces when users return a new sample rate from the `tracesSampler` in backend services, we propagate the generated random value used by the SDK for computing the sampling decision instead of sampling with a different random value in every service. Therefore, across a trace every SDK uses the same random value.

Without a `tracesSampler` callback, an SDK fully inherits sampling decisions for propagated traces. In this case, the presence of `sample_rand` in the DSC has no effect on the sampling decision. However, this behavior is subject to change in the future.

The random value is set according to the following rules:

> 1. When an SDK starts a new trace, `sample_rand` is always set to a truly random number in range `[0, 1]`. This explicitly includes traces that are not sampled, as well as when the `tracesSampleRate` is set to `0.0` or `1.0`.

Author: ```suggestion
1. When an SDK starts a new trace, `sample_rand` is always set to a random number in range `[0, 1]`. This explicitly includes traces that are not sampled, as well as when the `tracesSampleRate` is set to `0.0` or `1.0`.
```

> 2. It is _recommended_ to generate the random number deterministically using the trace ID as seed or source of randomness. The exact method by which the random number is created is implementation defined and may vary between SDK implementations.

Reviewer 4: ```suggestion
### Propagated Random Value

To improve the likelihood of capturing complete traces when backend services use a custom sample rate via `tracesSampler`, the SDK propagates the same random value used for sampling decisions across all services in a trace. This ensures consistent sampling decisions instead of generating a new random value for each service.

If no `tracesSampler` callback is used, the SDK fully inherits sampling decisions for propagated traces, and the presence of `sample_rand` in the DSC doesn’t affect the decision. However, this behavior may change in the future.

The random value is set according to the following rules:

1. When an SDK starts a new trace, `sample_rand` is always set to a random number in the range of `[0, 1]`. This explicitly includes traces that aren't sampled, as well as when the `tracesSampleRate` is set to `0.0` or `1.0`.
2. Generating the random number deterministically, using the trace ID as a seed or source of randomness is recommended. The method for creating this random value may vary between SDKs.
```

> 2. It is _recommended_ to generate the random number deterministically using the trace ID as seed or source of randomness. The exact method by which the random number is created is implementation defined and may vary between SDK implementations.

Author: FMI, explain the why

1. When the `tracesSampler` is invoked, this also applies to the return value of traces sampler. That is, `trace["sentry-sample_rand"] < tracesSampler(context)`
2. Otherwise, when the SDK is the head of a trace, this also applies to sample decisions based on `tracesSampleRate`. That is, `trace["sentry-sample_rand"] < config.tracesSampleRate`
3. There is no more direct comparison with `math.random()` during the sampling process.

When using a `tracesSampler`, the correct way to inherit parent sampling decisions is now to return the parent's sample rate instead of the decision as float (1.0). This way, Sentry can still extrapolate counts correctly.

```js
tracesSampler: ({ name, parentSampleRate }) => {
  // Inherit the trace parent's sample rate if there is one. Sampling is deterministic
  // for one trace, i.e. if the parent was sampled, we will be sampled too at the same
  // rate.
  if (typeof parentSampleRate === "number") {
    return parentSampleRate;
  }

  // Else, use default sample rate (replacing tracesSampleRate).
  return 0.5;
},

> ```

Reviewer 4: ```suggestion
   3. If `sampled` is missing the `sample_rate`, generate a random number in the range of `[0, 1]`, just like for a new trace.

The SDK should always use the stored random number (`sentry-sample_rand`) for sampling decisions and should no longer rely on `math.random()` or similar functions in tracing code:

1. When the `tracesSampler` is invoked, this also applies to `trace["sentry-sample_rand"] < tracesSampler(context)`, the return value of the traces sampler. 
2. If the SDK is starting a new trace, it uses the random number for decisions based on `tracesSampleRate``: `trace["sentry-sample_rand"] < config.tracesSampleRate`.
3. There are no more direct comparisons with `math.random()` during the sampling process.

When using a `tracesSampler`, the proper way to inherit a parent's sampling decision is to return the parent's sample rate, instead of leaving the decision as a float (for example, 1.0). This way, Sentry can still extrapolate counts correctly.

```js
tracesSampler: ({ name, parentSampleRate }) => {
  // Inherit the trace parent's sample rate if there is one. Sampling is deterministic
  // for one trace, i.e. if the parent was sampled, other traces will be sampled at the same
  // rate.
  if (typeof parentSampleRate === "number") {
    return parentSampleRate;
  }

  // Else, use default sample rate (replacing tracesSampleRate).
  return 0.5;
},
```
```

A transaction's sampling decision should be passed to all of its children, including across service boundaries. This can be accomplished in the `startChild` method for same-service children and using the `senry-trace` header for children in a different service.

### Propagated Random Value

To increase the chance of capturing complete traces when users return a new sample rate from the `tracesSampler` in backend services, we propagate the generated random value used by the SDK for computing the sampling decision instead of sampling with a different random value in every service. Therefore, across a trace every SDK uses the same random value.

> Without a `tracesSampler` callback, an SDK fully inherits sampling decisions for propagated traces. In this case, the presence of `sample_rand` in the DSC has no effect on the sampling decision. However, this behavior is subject to change in the future.

Reviewer 3: Maybe a stupid question, but what's the DSC? Dynamic Sampling Callback?

Author: https://develop.sentry.dev/sdk/telemetry/traces/dynamic-sampling-context/

Reviewer 3: 🤦 thanks

3. On incoming traces, an SDK assumes the `sample_rand` value along with the rest of the DSC, overriding an existing value if needed.
4. If `sample_rand` is missing on an incoming trace, the SDK creates and from now on propagates a new random number on-the-fly, based on the following rules:
   1. If `sample_rate` and `sampled` are propgated, create `sample_rand` so that it adheres to the invariant. This means, for a decision of `True` generate a random number in half-open range `[0, rate)` and for a decision of `False` generate a random number in range `[rate, 1]`.
   2. If the `sampled` is missing `sample_rate`, generate a random number in the range `[0, 1]` like for a new trace.

The SDK should exclusively use the stored random number and no longer use `math.random()` or similar anywhere else in the tracing related code base:

1. When the `tracesSampler` is invoked, this also applies to the return value of traces sampler. That is, `trace["sentry-sample_rand"] < tracesSampler(context)`
2. Otherwise, when the SDK is the head of a trace, this also applies to sample decisions based on `tracesSampleRate`. That is, `trace["sentry-sample_rand"] < config.tracesSampleRate`

> 3. There is no more direct comparison with `math.random()` during the sampling process.

Reviewer 3: I would really appreciate adding some examples, like we have above for `strictTracePropagation,` to see these rules in action. It would help me understand better how that actually works.

Author: Sure!
